test/solvers/functional/functional_test.py

```diff
@@ -0,0 +1,602 @@
+    wrong_space = odl.uniform_discr(1, 2, 10)
+    func_wrong_domain = odl.solvers.L1Norm(wrong_space)
+    with pytest.raises(TypeError):
+        func1 + func_wrong_domain
+
+    func_sum = func1 + func2
+    x = noise_element(space)
+
+    # Test evaluation of the functionals
+    expected_result = func1(x) + func2(x)
+    assert almost_equal(func_sum(x), expected_result, places=PLACES)
+
+    # Test for the gradient
+    expected_result = func1.gradient(x) + func2.gradient(x)
+    assert all_almost_equal(func_sum.gradient(x), expected_result,
+                            places=PLACES)
+
+    # Test that prox and convex conjugate is not known
+    with pytest.raises(NotImplementedError):
+        func_sum.proximal()
+    with pytest.raises(NotImplementedError):
+        func_sum.convex_conj()
+
+
+def test_functional_plus_scalar():
+    """Test for sum of functioanl and scalar."""
+    space = odl.uniform_discr(0, 1, 10)
+
+    func = odl.solvers.L2NormSquared(space)
+    scalar = np.random.randn()
+
+    # Test for scalar not in the field (field of unifor_discr is RealNumbers)
+    complex_scalar = np.random.randn() + np.random.randn() * 1j
+    with pytest.raises(TypeError):
+        func + complex_scalar
+
+    func_scalar_sum = func + scalar
+    x = noise_element(space)
+    p = noise_element(space)
+
+    # Test for evaluation
+    expected_result = func(x) + scalar
+    assert almost_equal(func_scalar_sum(x), expected_result, places=PLACES)
+
+    # Test for derivative and gradient
+    grad_x = func.gradient(x)
+    assert all_almost_equal(func_scalar_sum.gradient(x), grad_x, places=PLACES)
+
+    expected_result = grad_x.inner(p)
+    assert almost_equal(func_scalar_sum.derivative(x)(p), expected_result,
+                        places=PLACES)
+
+    # Test proximal operator
+    sigma = np.random.rand()
+    expected_result = func.proximal(sigma)(x)
+    assert all_almost_equal(func_scalar_sum.proximal(sigma)(x),
+                            expected_result, places=PLACES)
+
+    # Test convex conjugate functional
+    cc_func = func.convex_conj
+    cc_func_scalar_sum = func_scalar_sum.convex_conj
+    expected_result = cc_func(x) - scalar
+    assert almost_equal(cc_func_scalar_sum(x), expected_result, places=PLACES)
+
+    expected_result = cc_func.gradient(x)
+    assert all_almost_equal(cc_func_scalar_sum.gradient(x), expected_result,
+                            places=PLACES)
+
+
+def test_translation_of_functional():
+    """Test for the translation of a functional: (f(. - y))^*"""
+    space = odl.uniform_discr(0, 1, 10)
+
+    # The translation; an element in the domain
+    translation = noise_element(space)
+
+    # Creating the functional ||x||_2^2
+    test_functional = odl.solvers.L2NormSquared(space)
+
+    # Testing that translation belonging to the wrong space gives TypeError
+    wrong_space = odl.uniform_discr(1, 2, 10)
+    wrong_translation = noise_element(wrong_space)
```

Comment:
Take the zero element to ensure that there is nothing else causing a problem `wrong_translation = wrong_space.zero()`